src/main/java/picocli/CommandLine.java

```diff
@@ -2009,13 +2009,45 @@ private static int handleUnhandled(Exception ex, CommandLine cmd, int defaultExi
      */
     private static String exceptionToColorString(Exception ex, Help.ColorScheme existingColorScheme){
         Help.ColorScheme colorScheme = new Help.ColorScheme.Builder(existingColorScheme).applySystemProperties().build();
-        String newLine = System.getProperty("line.separator");
-        StringBuilder stringBuilder = new StringBuilder();
-        stringBuilder.append(colorScheme.errorText(ex.toString())).append(newLine);
-        for (StackTraceElement traceElement : ex.getStackTrace()) {
-            stringBuilder.append(colorScheme.stackTraceText("\tat " + traceElement)).append(newLine);
+        StringWriter stringWriter = new ColorSchemedStringWriter(colorScheme);
+        ex.printStackTrace(new PrintWriter(stringWriter));
+        return stringWriter.toString();
+    }
+
+    /**
+     * Extends StringWriter to use ColorScheme. Allows separating
+     * exception messages from stack traces by intercepting write method.
+     */
+    static class ColorSchemedStringWriter extends StringWriter {
+
+        Help.ColorScheme colorScheme;
+
+        public ColorSchemedStringWriter(Help.ColorScheme colorScheme) {
+            this.colorScheme = colorScheme;
+        }
+
+        public ColorSchemedStringWriter(int initialSize, Help.ColorScheme colorScheme) {
+            super(initialSize);
+            this.colorScheme = colorScheme;
         }
-        return stringBuilder.toString();
+
+        @Override
+        public void write(String str, int off, int len) {
+            if(str.startsWith("\t")){
+                // that's stacktrace
+                super.write(Style.on(colorScheme.stackTraceStyles().toArray(new IStyle[0])));
```

Comment:
It is not a good idea to unconditionally emit ANSI escape codes.
The overall idea is a good one, but it is better to construct `Ansi.Text` objects for the existing `Ansi` value.
The `ColorScheme.apply` method is useful here:

```java
@Override
public void write(String str, int off, int len) {
    List<IStyle> styles = str.startsWith("\t") ? colorScheme.stackTraceStyles() : colorScheme.errorStyles();
    super.write(colorScheme.apply(str.substring(off, len), styles).toString());
}
```

Reply:
Thanks for the advice! That seems much better!